File: compiler/bench/bubblesort.fs
\ expected: 499500
\ Bubble sort benchmark - sort 1000 elements descending, sum result

1000 constant SIZE
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: swap-arr ( i j -- )
  over arr@ over arr@
  rot arr! swap arr! ;

: bubble-sort ( n -- )
  dup 0 do
    dup 1- i - 0 do
      i arr@ i 1+ arr@ > if
        i i 1+ swap-arr
      then
    loop
  loop drop ;

: init-arr ( -- )
  SIZE 0 do SIZE i - 1- i arr! loop ;

: sum-arr ( -- n )
  0 SIZE 0 do i arr@ + loop ;

: main ( -- )
  init-arr
  SIZE bubble-sort
  sum-arr . cr ;

File: compiler/bench/bubblesort.c
// expected: 499500
// Bubble sort benchmark - sort 1000 elements descending, sum result

#include <stdio.h>

#define SIZE 1000

int arr[SIZE];

void swap_arr(int i, int j) {
    int tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
}

void bubble_sort(int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n - 1 - i; j++) {
            if (arr[j] > arr[j + 1]) {
                swap_arr(j, j + 1);
            }
        }
    }
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = SIZE - i - 1;
    }
}

long sum_arr(void) {
    long sum = 0;
    for (int i = 0; i < SIZE; i++) {
        sum += arr[i];
    }
    return sum;
}

int main(void) {
    init_arr();
    bubble_sort(SIZE);
    printf("%ld\n", sum_arr());
    return 0;
}
